OpenMEEG/include/analytics.h

```diff
@@ -162,56 +169,116 @@ namespace OpenMEEG {
+            r21 = p1-p2;
+            r2  = p2;
+
+            const double norm_r20_squared = r20.norm2();
+            const double norm_r21_squared = r21.norm2();
+            const double prod_r20_r21     = dotprod(r20,r21);
+
+            const double det = norm_r20_squared*norm_r21_squared-sqr(prod_r20_r21);
+            const double invdet = 1.0/det;
+
+            c11 = norm_r20_squared*invdet;
+            c22 = norm_r21_squared*invdet;
+            c12 = prod_r20_r21*invdet;
+        }
+
+        Vect3 operator()(const Vect3& r) const {
+
+            const Vect3& u = r-r2;
+            const double b1 = dotprod(r20,u);
+            const double b2 = dotprod(r21,u);
+            const double l = b1*c22-b2*c12;
+            const double m = b2*c11-b1*c12;
+            return Vect3(l,m,1-l-m);
         }
 
+    private:
+
+        double c11,c12,c22;
+        Vect3  r20,r21,r2;
+    };
+
+    class OPENMEEG_EXPORT analyticMonopolePotDer {
+    public:
+
+        analyticMonopolePotDer(const Monopole& monop,const Triangle& T):
+            barycentric_coords(T),triangle(T),monopole(monop)
+        { }
+
         Vect3 f(const Vect3& r) const {
-            Vect3 P1part(dotprod(H0p0DivNorm2,r-H0),dotprod(H1p1DivNorm2,r-H1),dotprod(H2p2DivNorm2,r-H2));
 
-            // B = n.grad_x(A) with grad_x(A)= q/||^3 - 3r(q.r)/||^5
+            // B = n.grad_x(A) with grad_x(A)= q x/|x|^3, with x = r-r0
+
+            const Vect3& x      = r-monopole.position();
+            const double xnrm2  = x.norm2();
+            const Vect3& n      = triangle.normal();
+			const double EMpart = monopole.charge()*dotprod(n,x)/(xnrm2*sqrt(xnrm2));
+            const Vect3& P1term = barycentric_coords(r); // P1 function values are just the barycentric coordinates of r.
+
+            return EMpart*P1term;
+        }
+
+    private:
+
+        const BarycentricCoordinates barycentric_coords;
+
+        const Triangle& triangle;
+        const Monopole& monopole;
+    };
+
+    class OPENMEEG_EXPORT analyticDipPotDer {
+    public:
+
+        analyticDipPotDer(const Dipole& dip,const Triangle& T):
+            barycentric_coords(T),triangle(T),dipole(dip)
+        { }
+
+        Vect3 f(const Vect3& r) const {
+
+            // B = n.grad_x(A) with grad_x(A)= q/|x|^3 - 3r(q.x)/|x|^5 with x = r-r0
 
             const Vect3& x         = r-dipole.position();
             const double inv_xnrm2 = 1.0/x.norm2();
-            const double EMpart = dotprod(n,dipole.moment()-3*dotprod(dipole.moment(),x)*x*inv_xnrm2)*(inv_xnrm2*sqrt(inv_xnrm2));
+            const Vect3& n         = triangle.normal();
+            const double EMpart    = dotprod(n,dipole.moment()-(3*dotprod(dipole.moment(),x)*inv_xnrm2)*x)*(inv_xnrm2*sqrt(inv_xnrm2));
+            const Vect3& P1term    = barycentric_coords(r); // P1 function values are just the barycentric coordinates of r.
 
-            return -EMpart*P1part; // RK: why - sign ?
+            return -EMpart*P1term; // RK: why - sign ?
```

Comment:
I agree with this - very old - comment.. I saw that in the analyticMonopolePotDer follows this sign change (at line 246 there should be a minus sign - per the mathematical formula  - which is not there) . Did you do it to match the sign here, or have you validated by some other mean (comparison with some ground truth, or analysis of the code at the upper level) that this sign change is needed.

Reply:
I do not really know where the sign question comes from because I do not know your derivations but I am sure it is correct that the monopole has opposite sign in this sense here. The monopole derivative is -1/4pi q@r/|r|^3 while the dipole is 1/4pi (q/|r|^3-3q@r/|r|^5*r)